### PR TITLE
fix: Assemblies enumerated multiple times

### PR DESCRIPTION
**Pull request by RussKie**
Whenever you select "Help File", "Components", "Plug-Ins" or "Transform Args" project property tabs the UI freezes and there is a noticable delay in loading the tab.
It appears the folders are traversed over and over again, the files are loaded over and over again.
The traversed paths are not recorded, which causes them to be traversed again whenever the `ComponentUtilities.CreateComponentContainer(...)` method is called.

The fix addressed the issue by making catalog and searched folders objects static.

### CODE REVIEW

File: SHFB/Source/SandcastleCore/ComponentUtilities.cs [hunk 1]
```diff
@@ -205,111 +209,138 @@ private static void SetPaths()
...
+        ///     folder.</param>
         /// <remarks>It is done this way to prevent a single assembly that would normally be discovered via a
         /// directory catalog from preventing all assemblies from loading if it cannot be examined when the parts
         /// are composed (i.e. trying to load a Windows Store assembly on Windows 7).</remarks>
-        private static void AddAssemblyCatalogs(AggregateCatalog catalog, string folder,
-          HashSet<string> searchedFolders, bool includeSubfolders)
+        private static void AddAssemblyCatalogs(string folder, bool includeSubfolders)
         {
-            if(!String.IsNullOrWhiteSpace(folder) && Directory.Exists(folder) && !searchedFolders.Contains(folder))
+            if (String.IsNullOrWhiteSpace(folder) || !Directory.Exists(folder) || SearchedFolders.Contains(folder))
```
Comment (RussKie) [inline SHFB/Source/SandcastleCore/ComponentUtilities.cs]: Inverted the condition to simplify the code.

**Comment by RussKie** [comment]
A better way to enumerate and load all assemblies into the catalog is do it asynchronously.
I have tested it and async is significantly faster. Can push it as a separate PR.

**Comment by EWSoftware** [comment]
There's a problem with this though and is why it's not done statically already.  The composition container is disposable.  When it gets disposed of, it disposes of the catalog.  At that point you're reusing a disposed catalog which is going to cause problems.  I can perhaps see doing it asynchronously without it being static.

**Comment by RussKie** [comment]
The composition container is indeed disposable, but being disposed it doesn't dispose the catalog it has been initialised with.
I have tested it in both debug and release modes.

**Comment by EWSoftware** [comment]
I checked with Reflector and it does disposed of its catalogs.  All catalogs implement IDisposable and will be disposed of.  It may not have failed when you tested it but that's not to say it won't eventually.

**Comment by RussKie** [comment]
I still disagree. I have checked the code in Reflector as well and I could not observe anything which would dispose the catalog, only its wrapper - CatalogExportProvider.
From the API design perspective, a class must only dispose objects it creates, any dependency passed in should not be disposed automatially.

There's also a SO thread confirming my observations - http://stackoverflow.com/questions/5675881/does-the-mef-container-dispose-dispose-added-catalogs

**Comment by EWSoftware** [comment]
Regardless, I think I'd prefer the async changes over the static implementation.

**Comment by RussKie** [comment]
Without statics (or other means to persist the catalog) you will be loading assemblies over and over again. Which IMHO is a poor UX. 
The async implementation I have leverages off the statics in this PR.

Ultimately it is your decision, feel free to decline.

**Comment by EWSoftware** [comment]
Disposal aside, the given implementation accumulates locations and assemblies across all projects and thus project-specific components in one project could be seen in another that didn't contain them.  The primary issue here is one of UI responsiveness.  As such, I've implemented an asynchronous cache that the property pages now use to improve their responsiveness.  This solves the issue at hand and keeps components local to the projects in which they are used.

**Comment by RussKie** [comment]
Awesome, I'll check it out.

**Pull request closed without merging**